diff --git a/bfd/dwarf2.c b/bfd/dwarf2.c
index 767e9731199..c7561cdeeec 100644
--- a/bfd/dwarf2.c
+++ b/bfd/dwarf2.c
@@ -1182,7 +1182,7 @@ read_attribute_value (struct attribute *  attr,
     case DW_FORM_ref_addr:
       /* DW_FORM_ref_addr is an address in DWARF2, and an offset in
 	 DWARF3.  */
-      if (unit->version == 3 || unit->version == 4)
+      if (unit->version >= 3)
 	{
 	  if (unit->offset_size == 4)
 	    attr->u.val = read_4_bytes (unit->abfd, info_ptr, info_ptr_end);
-- 
2.31.1

